packages/samples/headless-react/src/pages/DependentFacetPage.tsx

```diff
@@ -61,87 +33,47 @@ export class DependentFacetPage extends Component<
     this.fileTypeFacet = buildFacet(this.engine, {
       options: {field: 'filetype'},
     });
-    this.state = {currentExample: 'SingleParentSingleDependent'};
+  }
+
+  establishDependency(
+    dependentFacet: HeadlessFacet,
+    parentFacet: HeadlessFacet
+  ) {
+    this.dependenciesManagers.push(
```

Comment:
The second we start introducing "facet dependency" in Headless, then by the very definition of what this means, we must properly support and expose building blocks that correctly represent what users will want to do. 

If what we expose does not exactly represent what our users might want to achieve, then it means we must start adding more options, more corner case so that it works for "everyone".

The fact that you need to use/build multiple facet dependency manager, and that user can only do "one parent to one child, and then loop multiple time and push that in an array" is * not * particularly intuitive. 

As a user, I would think that there's something I must not be doing correctly, or that Headless does not support what I want to achieve.
I would expect it to accept some kind of complex tree as input upfront.

That's the original problem I wanted to solve by not having any code in Headless related to this feature.

Similar to "give a man a fish VS teach a man to fish" :)

The helper function approach can work and we can keep it.

Just that it could also be done in a "sample/look at this demo/copy and adapt this generic function".

Reply:
> The second we start introducing "facet dependency" in Headless, then by the very definition of what this means, we must properly support and expose building blocks that correctly represent what users will want to do.
>
> If what we expose does not exactly represent what our users might want to achieve, then it means we must start adding more options, more corner case so that it works for "everyone".
> [...]
> That's the original problem I wanted to solve by not having any code in Headless related to this feature.
> [...]
> it could also be done in a "sample/look at this demo/copy and adapt this generic function".

Honestly, after seeing all the potential issues with the url manager and with analytics that dependsOn can have, I think we should make the commitment of providing a helper/controller. Otherwise, I'm very worried the amount of support cases & broken analytics we would get would give us a greater amount of work than the helper/controller would.

> The fact that you need to use/build multiple facet dependency manager, and that user can only do "one parent to one child, and then loop multiple time and push that in an array to achieve complex dependencies" is * not * particularly intuitive.
>
> As a user, I would think that there's something I must not be doing correctly, or that Headless does not support what I want to achieve.
I would expect it to accept some kind of complex tree as input upfront.

I'm not sure I fully agree, maybe what's wrong here is partly the name of the controller. In this PR, the controller represents the dependencies of one facet. To me that's fairly intuitive, I'd like to hear more opinions to know if that may just be me. If we collapse everything into a single helper/controller, then managing something like a facet being removed/replaced becomes complex.